pkg/api/vm/handler.go

```diff
@@ -1164,6 +1181,167 @@ func (h *vmActionHandler) removeVolume(ctx context.Context, namespace, name stri
+		return err
+	}
+
+	ok, err := isVmNetworkHotpluggable(nad)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return fmt.Errorf("network %s is not hot-pluggable", input.NetworkName)
+	}
+
+	vmCopy := vm.DeepCopy()
+
+	newIface := kubevirtv1.Interface{
+		Name:                   input.InterfaceName,
+		MacAddress:             input.MacAddress,
+		Model:                  "virtio",
+		InterfaceBindingMethod: kubevirtv1.InterfaceBindingMethod{Bridge: &kubevirtv1.InterfaceBridge{}},
+	}
+	vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces = append(vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces, newIface)
+
+	newNetwork := kubevirtv1.Network{
+		Name:          input.InterfaceName,
+		NetworkSource: kubevirtv1.NetworkSource{Multus: &kubevirtv1.MultusNetwork{NetworkName: input.NetworkName}},
+	}
+	vmCopy.Spec.Template.Spec.Networks = append(vmCopy.Spec.Template.Spec.Networks, newNetwork)
+
+	_, err = h.vms.Update(vmCopy)
+	if err != nil {
+		return err
+	}
+	return h.migrate(ctx, namespace, name, "")
+}
+
+func (h *vmActionHandler) getHotunpluggableNetworks(vm *kubevirtv1.VirtualMachine) (map[string]struct{}, error) {
+	namespace := vm.ObjectMeta.Namespace
+	validNetworks := make(map[string]struct{})
+	for _, network := range vm.Spec.Template.Spec.Networks {
+		if network.Multus != nil {
+			nad, err := h.getVmNetwork(network.Multus.NetworkName, namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			ok, err := isVmNetworkHotpluggable(nad)
+			if err != nil {
+				return nil, err
+			}
+			if ok {
+				validNetworks[network.Name] = struct{}{}
+			}
+		}
+	}
+
+	return validNetworks, nil
+}
+
+func isInterfaceHotunpluggable(iface kubevirtv1.Interface, hotunpluggableNetworks map[string]struct{}) bool {
+	_, isNetworkHotunpluggable := hotunpluggableNetworks[iface.Name]
+	if iface.Model == "virtio" && iface.Bridge != nil && iface.State != kubevirtv1.InterfaceStateAbsent && isNetworkHotunpluggable {
+		return true
+	}
+	return false
+}
+
+// removeNic remove a hotplug NIC by its interface name
+func (h *vmActionHandler) removeNic(ctx context.Context, namespace, name string, input RemoveNicInput) error {
+	vm, err := h.vmCache.Get(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	hotunpluggableNetworks, err := h.getHotunpluggableNetworks(vm)
+	if err != nil {
+		return err
+	}
+
+	vmCopy := vm.DeepCopy()
+
+	found := false
+	interfaces := vmCopy.Spec.Template.Spec.Domain.Devices.Interfaces
+	for idx := range interfaces {
+		if interfaces[idx].Name == input.InterfaceName {
+			if isInterfaceHotunpluggable(interfaces[idx], hotunpluggableNetworks) {
+				interfaces[idx].State = kubevirtv1.InterfaceStateAbsent
+				found = true
+				break
+			}
+			return fmt.Errorf("interface %s is not hot-unpluggable", input.InterfaceName)
+		}
+	}
+
+	if !found {
+		return fmt.Errorf("interface %s doesn't exist", input.InterfaceName)
+	}
+
+	_, err = h.vms.Update(vmCopy)
+	if err != nil {
+		return err
+	}
+	return h.migrate(ctx, namespace, name, "")
```

Comment:
same as before, this may not be needed.

Reply:
https://kubevirt.io/user-guide/network/hotplug_interfaces/#migration-based-hotplug points out that it is not needed

Reply:
Oops, I miss that. Thanks for the note! I'll remove the explicit migration call and try it.